### ISSUE TITLE
Hidden package caching behavior breaks package exclusion

### ISSUE THREAD

**Issue by drdavella** [astropy/astropy-helpers#329]
The `get_package_info` command provides an optional parameter `exclude=`, which tells the `find_packages` routine in `setuptools` to ignore packages with names that match certain patterns. For example, a common use case would be to avoid installing tests:

```python
# setup.py
import ah_bootstrap
from setuptools import setup
from astropy_helpers.setup_helpers import get_package_info

...

package_info = get_package_info(exclude=['*.tests'])
setup(**package_info)
```

In many (if not all) cases, a project will also want to use custom commands provided by `astropy_helpers`. To do this, the code ends up looking like the following:

```python
# setup.py
import ah_bootstrap
from setuptools import setup
from astropy_helpers.setup_helpers import get_package_info, register_commands

...

release = False
cmdclassd = register_commands('foobar', '1.0.dev', release)
package_info = get_package_info(exclude=['*.tests'])
setup(cmdclass=cmdclassd, **package_info)
```

Here's the problem: when I call `register_commands`, it implicitly calls the underlying `find_packages` routine, but without any knowledge of the packages that I want to exclude. It then does me the favor of caching the packages that it found so that it doesn't have to search again. However, when I call `get_package_info`, it goes straight to the cached list of packages and does not filter against my requested exclusions.

I think I understand the issue well enough so I should be able to submit a PR with a fix sooner or later. Any input on the right approach is welcome, though. Right now it's not clear to me whether `register_commands` actually needs to know anything about the listed packages or not. If it doesn't, that would be an easy and clean fix.

**Comment by pllim** [comment]
Pinging @embray just in case he can comment on this... 🙏

**Comment by drdavella** [comment]
A possible solution could be to add an `exclude` parameter to `register_commands`, but it's not exactly intuitive and seems prone to user error. Another solution would be to disable caching when `find_packages` is called by `register_commands`, but that seems to defeat the purpose. It's not a hard fix, it's just a matter of finding the _right_ fix. :)

**Comment by drdavella** [comment]
This was closed by #331.
